packages/blocks/src/page-block/edgeless/components/edgeless-selected-rect.ts

```diff
@@ -0,0 +1,182 @@
+        const [x, y, w, h] = deserializeXYWH(getXYWH(s));
+        return { x, y, w, h };
+      }
+      return s;
+    });
+
+    return getCommonBound(bounds) as Bound;
+  }
+
+  private _onDragMove = (deltaBound: Bound) => {
+    const oldBound = this.modelBound;
+    const newBound = {
+      x: oldBound.x + deltaBound.x,
+      y: oldBound.y + deltaBound.y,
+      w: oldBound.w + deltaBound.w,
+      h: oldBound.h + deltaBound.h,
+    };
+
+    this.state.selected.forEach(element => {
+      const [ex, ey, ew, eh] = deserializeXYWH(getXYWH(element));
+      const x = (newBound.w * (ex - oldBound.x)) / oldBound.w + newBound.x;
+      const y = (newBound.h * (ey - oldBound.y)) / oldBound.h + newBound.y;
+      const w = (newBound.w * ew) / oldBound.w;
+      const h = (newBound.h * eh) / oldBound.h;
+
+      if (isTopLevelBlock(element)) {
+        let frameX = x;
+        let frameY = y;
+        let frameW = w;
+        let frameH = h;
+        // limit the width of the selected frame
+        if (frameW < FRAME_MIN_LENGTH) {
+          frameW = FRAME_MIN_LENGTH;
+          frameX = x;
+        }
+        // limit the height of the selected frame
+        if (frameH < FRAME_MIN_LENGTH) {
+          frameH = FRAME_MIN_LENGTH;
+          frameY = y;
+        }
+        const frameBlock = getBlockById<'div'>(element.id);
+        const frameContainer = frameBlock?.parentElement;
+        // first change container`s x/w directly for get frames real height
+        if (frameContainer) {
+          frameContainer.style.width = frameW + 'px';
+          frameContainer.style.translate = `translate(${frameX}px, ${frameY}px) scale(${this.zoom})`;
+        }
+        // reset the width of the container may trigger animation
+        requestAnimationFrame(() => {
+          // refresh xywh by model
+          if (!this._lock) {
+            this.page.captureSync();
+            this._lock = true;
+          }
+
+          const newXywh = JSON.stringify([
+            frameX,
+            frameY,
+            frameW,
+            (frameBlock?.getBoundingClientRect().height || 0) / this.zoom +
+              EDGELESS_BLOCK_CHILD_PADDING * 2,
+          ]);
+          this.page.updateBlock(element, { xywh: newXywh });
+        });
+      } else {
+        this.surface.setElementBound(element.id, { x, y, w, h });
+      }
+    });
+
+    this.requestUpdate();
+  };
+
+  private _onDragEnd = () => {
+    if (this._lock) {
+      this.page.captureSync();
+    }
+    this._lock = false;
+  };
+
+  firstUpdated() {
+    const { _disposables, slots } = this;
+    _disposables.add(slots.viewportUpdated.on(() => this.requestUpdate()));
+  }
+
+  disconnectedCallback() {
+    this._disposables.dispose();
+  }
+
+  render() {
+    if (this.state.selected.length === 0) return null;
+
+    const { page, state, surface, resizeMode, _resizeManager } = this;
+    const { active, selected } = state;
+    const selectedRect = getSelectedRect(selected, surface.viewport);
```

Comment:
I don't want to put calculating `selectedRect` in `selected rect` component. Because the modifier popup will reuse this information.

Reply:
Next feature🤣

Reply:
No worries, let's keep using minimal implementation, and refactor it by then :)